articles/testing-documentation-snippets/article.md

```diff
@@ -0,0 +1,173 @@
+
+
+## Writting documentation and snippets has to be easy
+
+The first task I have to make is to create a system easy to use for writing snippets. We use Metalsmith to generate our documentation website. It's a great tool and it uses Markdown to generate static HTML website. Our actual documentation use snippets written directly in Markdown. First, step was to take out the snippets and put them in a separate file with the corresponding extension for each language (mySnippet.js, mySnippet.go …). And after that, I just had to develop a tiny Metalsmith plugin to insert snippet code in markdown. With this plugin, when we have to write a snippet, we just need to add a special tag in the markdown: **“[snippet=feature_name]”** and the plugin automatically inject the content of all the snippets files contained in the **snippet** folder to the generated HTML.
+
+Here we are, we have a working markdown templating system with real source code files as  snippets examples, able to generate HTML.
+
+![schema1](./images/schema1.png)
+
+## Executing these snippets
+
+By definition, snippets are just a piece of code and they aren't often able to be played without a context (other pieces of code). We imagined a template system in which snippets content will be injected.
+
+Here, an example of a snippet in javascript
+
+```js
+//Connects to Kuzzle
+kuzzle
+  .connect()
+  .then(() => {
+    console.log('Successfully connected');
+  })
+  .catch(error => {
+    console.error(error.message);
+  });
+```
+
+And here a code template
+
+```js
+// load the Kuzzle SDK module
+const { Kuzzle } = require('kuzzle-sdk');
+
+// instantiate a Kuzzle client
+const kuzzle = new Kuzzle('websocket', {
+  host: 'kuzzle',
+  autoReconnect: false
+});
+
+// snippet will be injected here
+[snippet-code]
+
+```
+
+The tag **[snippet-code]** in the template will be replaced by the snippet and will form a full context file which can be executed.
+
+![schema2](./images/schema2.png)
+
+
+At this point, we have many source files, in many languages. First version of this new documentation system can support snippets test for Javascript, Java, C++ and Go. Some languages are interpreted, others compiled. All of them need their dedicated toolchain !
+
+![use docker](./images/use-docker.jpg)
+
+First thing I do is to create a DockerFile for each language implemented with all the toolchain to compile/interpret generated file. Also, we decided to add a linter in each container for each language to be sure to have a correct syntax and minimize source of errors (you can see an exemple of a dockerfile [here](https://hub.docker.com/r/kuzzleio/documentation-v2/~/dockerfile/)).
+
+At this point, we had some docker containers able to run code for their dedicated language and things seem to work perfectly!
```

Comment:
At this point, we could run each snippet in its dedicated container and things seem to work perfectly!